File: db/routers/secrets.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
import typing as t

from db import crud, schemas
import db.models.inventory_models as models
from db.utils.crypto import encrypt_secret, secrets_key_status
from db.utils.audit import log_event
from ..dependencies import get_db

router = APIRouter()


@router.get("/status")
def secrets_status() -> t.Any:
    configured, message = secrets_key_status()
    return {"configured": configured, "message": message}


@router.get("", response_model=list[schemas.AppSecretMeta])
def list_secrets(db: Session = Depends(get_db)) -> t.Any:
    """
    List secret metadata only (never returns plaintext secret values).
    """
    secrets = crud.secrets.get_all(db)
    # Ensure response includes `is_set` without exposing encrypted_value
    return [
        schemas.AppSecretMeta(
            name=s.name,
            is_active=s.is_active,
            is_set=True,
            created_at=s.created_at,
            updated_at=s.updated_at,
        )
        for s in secrets
    ]


@router.put("/{name}", response_model=schemas.AppSecretMeta)
def set_secret(
    name: str, body: schemas.AppSecretSet, db: Session = Depends(get_db)
) -> t.Any:
    """
    Upsert a secret value by name (stores encrypted at rest).
    """
    if not name or not name.strip():
        raise HTTPException(status_code=400, detail="Secret name is required")
    if body.value is None or body.value == "":
        raise HTTPException(status_code=400, detail="Secret value is required")

    existing = crud.secrets.get_by(db, obj_in={"name": name})
    try:
        encrypted_value = encrypt_secret(body.value)
    except RuntimeError as e:
        # Most common: GALAGO_SECRETS_KEY missing/invalid
        raise HTTPException(status_code=500, detail=str(e))

    if existing is None:
        created = models.AppSecret(
            name=name,
            encrypted_value=encrypted_value,
            is_active=body.is_active,
        )
        db.add(created)
        db.commit()
        db.refresh(created)
        s = created
    else:
        existing.encrypted_value = encrypted_value
        existing.is_active = body.is_active
        db.add(existing)
        db.commit()
        db.refresh(existing)
        s = existing

    log_event(
        db,
        action="secrets.set",
        target_type="secret",
        target_name=name,
        details={"is_active": body.is_active},
    )

    return schemas.AppSecretMeta(
        name=s.name,
        is_active=s.is_active,
        is_set=True,
        created_at=s.created_at,
        updated_at=s.updated_at,
    )


@router.delete("/{name}")
def clear_secret(name: str, db: Session = Depends(get_db)) -> t.Any:
    existing = crud.secrets.get_by(db, obj_in={"name": name})
    if existing is None:
        # Idempotent delete
        return {"message": "Secret not found"}
    db.delete(existing)
    db.commit()
    log_event(
        db, action="secrets.clear", target_type="secret", target_name=name, details=None
    )
    return {"message": "Secret cleared"}

File: controller/server/routers/secrets.ts
import { procedure, router } from "@/server/trpc";
import { db } from "@/db/client";
import { appSecrets } from "@/db/schema";
import { eq } from "drizzle-orm";
import { z } from "zod";
import { encryptSecret, secretsKeyStatus } from "@/server/utils/secretsCrypto";
import { logAuditEvent } from "@/server/utils/auditLog";

export interface AppSecretMeta {
  name: string;
  is_active: boolean;
  is_set: boolean;
  created_at?: Date;
  updated_at?: Date;
}

export interface SecretsStatus {
  configured: boolean;
  message?: string | null;
}

export const secretsRouter = router({
  status: procedure.query(async () => {
    return secretsKeyStatus();
  }),

  getAll: procedure.query(async () => {
    const rows = await db.select().from(appSecrets);
    return rows.map(
      (r): AppSecretMeta => ({
        name: r.name,
        is_active: Boolean(r.isActive),
        is_set: true,
        created_at: r.createdAt ? new Date(r.createdAt) : undefined,
        updated_at: r.updatedAt ? new Date(r.updatedAt) : undefined,
      }),
    );
  }),

  set: procedure
    .input(
      z.object({
        name: z.string().min(1),
        value: z.string().min(1),
        is_active: z.boolean().optional(),
      }),
    )
    .mutation(async ({ input }) => {
      const { name, value } = input;
      const isActive = input.is_active ?? true;
      const encrypted = encryptSecret(value);

      const existing = await db.select().from(appSecrets).where(eq(appSecrets.name, name)).limit(1);
      if (existing.length > 0) {
        const updated = await db
          .update(appSecrets)
          .set({ encryptedValue: encrypted, isActive })
          .where(eq(appSecrets.name, name))
          .returning();
        const r = updated[0];
        await logAuditEvent({
          action: "secrets.set",
          targetType: "secret",
          targetName: name,
          details: { is_active: isActive },
        });
        return {
          name: r.name,
          is_active: Boolean(r.isActive),
          is_set: true,
          created_at: r.createdAt ? new Date(r.createdAt) : undefined,
          updated_at: r.updatedAt ? new Date(r.updatedAt) : undefined,
        } satisfies AppSecretMeta;
      }

      const created = await db
        .insert(appSecrets)
        .values({ name, encryptedValue: encrypted, isActive })
        .returning();
      const r = created[0];
      await logAuditEvent({
        action: "secrets.set",
        targetType: "secret",
        targetName: name,
        details: { is_active: isActive },
      });
      return {
        name: r.name,
        is_active: Boolean(r.isActive),
        is_set: true,
        created_at: r.createdAt ? new Date(r.createdAt) : undefined,
        updated_at: r.updatedAt ? new Date(r.updatedAt) : undefined,
      } satisfies AppSecretMeta;
    }),

  clear: procedure.input(z.string()).mutation(async ({ input }) => {
    const name = input;
    await db.delete(appSecrets).where(eq(appSecrets.name, name));
    await logAuditEvent({
      action: "secrets.clear",
      targetType: "secret",
      targetName: name,
      details: null,
    });
    return { message: "Secret cleared" };
  }),
});
